presto-orc/src/main/java/com/facebook/presto/orc/reader/DecimalStreamReader.java

```diff
@@ -183,47 +125,60 @@ private void readStreamsData(DecimalType decimalType)
                 }
 
                 if (decimalType.isShort()) {
-                    decimalStream.nextLongVector(nextBatchSize, shortDecimalVector, nullVector);
+                    decimalStream.nextShortDecimalVector(nextBatchSize, builder, decimalType, scaleVector, nullVector);
                 }
                 else {
-                    decimalStream.nextBigIntegerVector(nextBatchSize, longDecimalVector, nullVector);
+                    decimalStream.nextLongDecimalVector(nextBatchSize, builder, decimalType, scaleVector, nullVector);
                 }
 
                 scaleStream.nextLongVector(nextBatchSize, scaleVector, nullVector);
             }
         }
+
+        readOffset = 0;
+        nextBatchSize = 0;
+
+        return builder.build();
     }
 
-    private Block buildDecimalsBlock(DecimalType decimalType)
-            throws OrcCorruptionException
+    private void openRowGroup()
```

Comment:
There was no reordering.
I just inlined the `readStreamsData`, while changing it a bit to match structure in other stream readers.

To make it easier to read I split this commit into two:
 1. changes (use `BlockBuilder` directly)
 2. inlines the method

Order of methods matches structure of other stream readers:
 * `prepareNextRead`
 * `readBlock`
 * `openRowGroup`
 * `startStripe`
 * `startRowGroup`

`DecimalStreamReader` has extra two private methods: 
 * `seekToOffset`
 * `allocateVectors`
but those seem to be in reasonable place in file